rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala

```diff
@@ -550,27 +550,30 @@ object ProcNormalizeMatcher {
         }
 
         val targetResult = normalizeMatch(b.proc_, input.copy(par = VectorPar()))
+        val outermostBundle = b.bundle_ match {
+          case _: BundleReadWrite => Bundle(targetResult.par, writeFlag = true, readFlag = true)
+          case _: BundleRead      => Bundle(targetResult.par, writeFlag = false, readFlag = true)
+          case _: BundleWrite     => Bundle(targetResult.par, writeFlag = true, readFlag = false)
+          case _: BundleEquiv     => Bundle(targetResult.par, writeFlag = false, readFlag = false)
+        }
 
-        //TODO(mateusz.gorski): implement actual collapsing/flattening of the bundles.
-        //Applicable for the case where bundle contains ONLY another bundle
-        def flatten(b: Bundle): Bundle = b
+        import BundleOps._
+
+        def flatten(b: Bundle): Bundle = {
+          import BundleOps._
+          b.body.get.singleBundle() match {
+            case Some(single) =>
+              b.merge(flatten(single)) // we need to collapse all the nested bundles
```

Comment:
You don't need to recurse here. A single level will suffice because you collapse as you go up.

Reply:
That is true. Didn't think of that. Thanks, that's generating a lot of necessary passes.